DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,696 @@
+> To clarify the difference, C strings have no abstraction whatever - they are a raw pointer to memory. However, they do have an invariant - they must point to a valid NUL-terminated string. Every place that constructs such a string must ensure it is valid, and every place that consumes it can rely on it.
+> OTOH, a safe, say, buffered reader needs abstraction but doesn't need protection - it does not hold any critical invariant, but may want to change its internal representation.
+
+[source](https://github.com/rust-lang/rfcs/issues/381#issuecomment-174955431)
+
+> This doesn't seem very useful to me. Within a module I would expect the authors to know what they're doing, and the unit-tests to save them when they do not.
+> For other users, you could simply introduce getters and setters, and functions/methods can already be marked unsafe.
+
+[source](https://github.com/rust-lang/rfcs/pull/80#issuecomment-43489000)
+
+Ultimately the proposal has not been accepted yet.
+The idea of using `private` instead of `@system` variables for D will be discussed in the [alternatives](#alternatives) section.
+More information about Rust's stand on unsafe functions can be found here:
+
+- [safe unsafe meaning](https://doc.rust-lang.org/nightly/nomicon/safe-unsafe-meaning.html)
+- [The scope of unsafe](https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html)
+
+## Description
+
+### Existing rules for `@system`
+
+Before the proposed changes, we quickly go over the relevant existing rules of what declarations can have the `@system` attribute:
+```D
+@system int w = 2; // compiles, does nothing
+@system enum int x = 3; // compiles, does nothing
+enum E {
+    @system x, // error: @system is not a valid attribute for enum members
+    y,
+}
+@system alias x = E; // compiles, does nothing
+@system template T() {} // compiles, does nothing
+
+void func(@system int x) // error: @system attribute for function parameter is not supported
+{
+    @system int x; // compiles, does nothing
+}
+template Temp(@system int x) {} // error: basic type expected, not @
+```
+In short, anything that can be marked `private` can also be marked `@system`.
+Additionally, local variables can be marked `@system` (while they can't be marked `private`).
+
+Any function attribute can be attached to a variable declaration, but they cannot be retrieved:
+```D
+@system @nogc pure nothrow int x;
+pragma(msg, __traits(getFunctionAttributes, x)); // Error: first argument is not a function
+pragma(msg, __traits(getAttributes, x)); // tuple()
+```
+
+### Unsafe types
+In the proposed changes I use the term "unsafe types".
+An unsafe type is a type which has underlying bit-patterns that risk causing memory corruption in `@safe` code.
+The prime example of an unsafe type is a pointer, but any type that contains a pointer is also unsafe:
```

Comment:
Why would these reference types be unsafe if they were allocated on the GC heap?

Reply:
I evidently didn't communicate things clearly, so I'm not sure how to answer this. Here are some key insights:
- The concepts in this DIP have nothing to do with memory allocation or lifetimes. In a world where freeing memory did not exist, the definitions still apply.
- A pointer is an unsafe type because if I where allowed to arbitrarily change the bits, I could construct a garbage pointer that can cause memory corruption in `@safe` code.
- By contrast, if I could arbitrarily change the bits of an `int` or `float`, it could get strange values and mess up the program, but because of bounds checks it is not sufficient for memory corruption in `@safe` code. There is not equivalent to a dereference operator on ints or floats, all their operators don't touch memory.
- With void initialization, union and array casting I can arbitrarily change bits of a variable regardless of type. That's why they are not allowed on pointers in `@safe` code, but they are on ints and floats.
- Using GC pointers is `@safe` precisely because pointers are an unsafe type: the language ensures pointer integrity. If pointers were a safe type, I could change it to something that is not a GC pointer and dereference it to corrupt memory.

I think the name "unsafe types" is confusing. The long name would be "types where the integrity of the underlying bits is necessary for memory safety".

Reply:
I think that all of what you just wrote belongs in the DIP itself.

Reply:
Yes, the DIP needs to be self-explanatory, and include all of this rationale so that it's understood in the proper context.